gateway/protocol.go

```diff
@@ -13,9 +13,12 @@ import (
 // Protocol defines the core functionality of a protocol from the perspective of a gateway.
 // The gateway expects a protocol to build and return a request context for a particular service ID.
 type Protocol interface {
+	// GetEndpoints returns a list of endpoints for a given service ID
+	GetEndpoints(serviceID protocol.ServiceID, httpReq *http.Request) (map[protocol.EndpointAddr]protocol.Endpoint, error)
+
 	// BuildRequestContext builds and returns a ProtocolRequestContext interface for handling a single service
 	// request, which matches the provided Service ID.
-	BuildRequestContext(protocol.ServiceID, *http.Request) (ProtocolRequestContext, error)
+	BuildRequestContext(protocol.ServiceID, map[protocol.EndpointAddr]protocol.Endpoint) (ProtocolRequestContext, error)
```

Comment:
Ditto re this method update: we are passing endpoints as "input" to the protocol: but it is the protocol's responsibility to know about and handle endpoints.